Improve inferability of `slicedim`

martinholters opened the pull request:
An attempt to fix #20141. Not perfect, but an improvement, and should do no harm.

E.g. with this PR:
```julia
julia> @code_warntype slicedim(ones(3), 1, 1)
# ...
  end::Union{Array{Float64,1},Float64}

julia> @code_warntype slicedim(ones(3), 1, 1:1)
# ...
  end::Array{Float64,1}

julia> @code_warntype slicedim(ones(3,3), 1, 1)
# ...
  end::Union{Array{Float64,1},Array{Float64,2},Float64}

julia> @code_warntype slicedim(ones(3,3), 1, 1:1)
# ...
  end::Array{Float64,2}

julia> @code_warntype slicedim(ones(3,3,3), 1, 1)
# ...
  end::Any

julia> @code_warntype slicedim(ones(3,3,3), 1, 1:1)
# ...
  end::Array{Float64,3}

julia> @code_warntype slicedim(ones(3,3,3,3,3,3), 1, 1:1)
# ...
  end::Array{Float64,6}

julia> @code_warntype slicedim(ones(3,3,3,3,3,3,3), 1, 1:1)
# ...
  end::Any
```
On master, they are all just `::Any`.

martinholters commented on base/abstractarraymath.jl:
```diff
@@ -91,6 +91,10 @@ imag{T<:Real}(x::AbstractArray{T}) = zero(x)
 
+_slice_inds(new_inds::Tuple, inds::Tuple, d, i) =
+    _slice_inds((new_inds..., length(new_inds) + 1 == d ? i : first(inds)), tail(inds), d, i)
+_slice_inds(new_inds::Tuple, inds::Tuple{}, d, i) = new_inds
```
Could also be
```julia
setindex(x::Tuple, n, v) = _setindex((), x, n, v)
_setindex(y::Tuple, x::Tuple, n, v) = _setindex((y..., length(y) + 1 == n ? v : first(x)), tail(x), n, v)
_setindex(y::Tuple, x::Tuple{}, n, v) = y
```
and live in tuple.jl. Or do we have something like this somewhere already?

timholy replied:
Seems like a good idea. You'd want the `v` to come before `n` in the argument order, though, for consistency with `setindex!`.

timholy commented on base/abstractarraymath.jl:
```diff
@@ -91,6 +91,10 @@ imag{T<:Real}(x::AbstractArray{T}) = zero(x)
 # index A[:,:,...,i,:,:,...] where "i" is in dimension "d"
 
+_slice_inds(new_inds::Tuple, inds::Tuple, d, i) =
+    _slice_inds((new_inds..., length(new_inds) + 1 == d ? i : first(inds)), tail(inds), d, i)
```
I'd consider `ifelse` here since otherwise you'll get one branch per dimension. Might also consider `@inline`ing this.

timholy reviewed (commented):
Maybe briefer would be `ntuple(j->ifelse(j==d,i,inds[d]), Val{N})`, but overall I like the idea of having an internal tuple-`setindex`, it makes it more readable.

I'd also point out that you may be able to get full inferrability of the output in most cases by manually allocating the output array and then using `copy!`. As long as the indices-tuple is homogenous, the indices-tuple of the output array differs by being of length `N-1` rather than `N`. This is the same type as `tail` would produce, so you could define

```jl
drop_dth{N,I}(inds::NTuple{N,I}, d) = _drop_dth(tail(inds), (), d, inds...)
_drop_dth(ref, out, d, ind, inds...) = length(out)==d+1 ? _drop_dth(ref, out, d, inds...) : _drop_dth(ref, (out..., ind), d, inds...)
_drop_dth{M,I}(::NTuple{M,I}, out, d)::NTuple{M,I} = out  # M = N-1
```

`drop_dth` is type-unstable at intermediate stages, but the output tuple has known type. Pass `out` as the indices-tuple to `similar`, and voila, the return array is inferrable.

(Might need a fallback for non-homogeneous indices tuples, but AFAIK there are no such arrays in existence.)

martinholters commented:
> the indices-tuple of the output array differs by being of length `N-1` rather than `N`

We support `d>N`, so the length of the output array indices tuple depends on the value of `d`.

timholy commented:
Seems almost worth having a separate code path for this case, e.g., require an extra argument or something. But that's not pretty, so maybe it's better to live with the type-instability.

martinholters committed:
Add `setindex` for tuples

martinholters commented:
Introduced `setindex(x::Tuple, v, i)` instead of `_slice_inds` and updated `slicedim` accordingly.

@timholy Does `ntuple` work for you the way you suggested? For the `(f, Val{N})` case, it seems it requires a type-stable `f`?

timholy commented:
Hmm, seems likely. This version looks fine to me.

martinholters commented:
Should `setindex` and/or `_setindex` be `@inline`? You probably have more experience whether/when this is beneficial.

timholy commented on base/tuple.jl:
```diff
@@ -18,6 +18,12 @@ getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
 
+# returns new tuple; N.B.: becomes no-op if i is out-of-bounds
+setindex(x::Tuple, v, i::Integer) = _setindex((), x, v, i::Integer)
+_setindex(y::Tuple, r::Tuple, v, i::Integer) =
```
This is the one to inline.

martinholters committed:
Improve inferability of `slicedim`

martinholters commented:
Added `@_inline_meta` and a test.

martinholters commented:
Oh, there is also a `slicedim` specialization for `BitArray` that could use a similar treatment. Will take a closer look tomorrow.

martinholters committed:
Remove `slicedim` specialization for `BitArray`

martinholters commented:
Interesting, the `BitArray` specialization seems to be significantly slower (orders of magnitude!) than the generic code. Hence I have just removed it. Could someone double-check whether my benchmark in [this gist](https://gist.github.com/martinholters/202bbe694f7b644f72170c8660e2ab9c) makes sense? (Note: I've run it against this branch, so by default there is no specialization.)

The specialized code also did not allow a "too large" dimension as long as indexing with 1, so could be made type-stable. But for consistency, probably it should either always be allowed or always disallowed.

martinholters commented:
Travis failure on Mac looks unrelated.

martinholters commented:
### Good news:
Doing some more benchmarking I am now pretty convinced that the generic code is significantly faster than the specialized code for `BitArray`, so removing it seems the right thing to do, except...
### Bad news:
#20233

If we want to preserve the current behavior, I could add back a specialization for `BitArray{1}`.

martinholters committed:
Add `slicedim(::BitVector, ::Integer, ::Integer)` to restore previous behavior

martinholters commented:
Added a specialization so that this PR only does what the title says and does not change any return types.

tkelman commented on base/bitarray.jl:
```diff
@@ -1278,39 +1278,20 @@ end
-            B[l] = A[j + index_offset]
-            l += 1
-        end
+# preserve some special behavior
```
was this tested very carefully before?

martinholters replied:
No, it's not tested at all. And I didn't feel like adding a test now because the desired behavior might actually be redefined shortly depending on how we resolve #20233.

StefanKarpinski commented:
Bump – this seems like a strict improvement.

martinholters commented:
Barring any objections, I'm going to merge this tomorrow.
